### PR TITLE
cert_validator: Add and populate the local address in extra validation context 

### PR DESCRIPTION
**Pull request by tyxia**
**Change**: Add and populate the local address in extra validation context so that custom cert_validator extension can use it if needed. This is only needed additional parameter at the moment and It is added as an optional field.

**Testing**: Modified the `timed_cert_validator` and added integration to validate the local address

### CODE REVIEW

File: source/extensions/transport_sockets/tls/context_impl.cc
Comment (danzh2010) [inline source/extensions/transport_sockets/tls/context_impl.cc:522]: can `localAddress()` return null here?

File: source/extensions/transport_sockets/tls/cert_validator/cert_validator.h
Comment (yanavlasov) [inline source/extensions/transport_sockets/tls/cert_validator/cert_validator.h:51]: Address::InstanceConstSharedPtr is a better choice here, since it is already a shared pointer. null would indicate it was not set. Posting a reference into an asynchronous routing is brittle as you need to be sure that lifetime of referent is sufficient for completion of cert validation (and all error cases).

**Review by yanavlasov — CHANGES_REQUESTED** [review]
/wait

**Review by ggreenway — CHANGES_REQUESTED** [review]
Why do you need the local address to verify the remote certificate?

What will this be used for? There isn't any usage of this data in the PR.

You should add an integration test to validate that this works as expected.

**Comment by tyxia** [comment]
> Why do you need the local address to verify the remote certificate?
> 
> What will this be used for? There isn't any usage of this data in the PR.
> 
> You should add an integration test to validate that this works as expected.

It is used for mTLS in our private extension.  Local address is required in the API communicating to the remote validation server. We will add the integration test for it internally and I added the comment here to clarify it a bit.

File: source/extensions/transport_sockets/tls/context_impl.cc
Comment (tyxia) [reply source/extensions/transport_sockets/tls/context_impl.cc]: I guess it will not be null but I have changed to shared ptr.

Comment (danzh2010) [reply source/extensions/transport_sockets/tls/context_impl.cc]: My point is that it is always safe to check nullptr before dereferencing a pointer.

**Comment by danzh2010** [comment]
+1 for a unit test for the self address you added.

File: source/extensions/transport_sockets/tls/context_impl.cc
Comment (tyxia) [reply source/extensions/transport_sockets/tls/context_impl.cc]: Yes, there is no dereference since the shared ptr is passed directly

**Comment by tyxia** [comment]
> +1 for a unit test for the self address you added.

I am wondering do you have the pointer to existing test example that will hit this function? I checked context_impl_test before. Whole function will not be hit.

**Comment by tyxia** [comment]
> timed_cert_validator

Thanks for example. Yes, that is also extension specific test, which i planned to test it in our extension test internally.

**Comment by danzh2010** [comment]
> > You can add a test cert validator to check the self address and config it for ssl_socket_test or some e2e tests. I just realize that cert validation logic is hard to test in unit test by the nature of boring SSL. An example will be: test/extensions/transport_sockets/tls/cert_validator/timed_cert_validator.h and its usage in integration tests
> 
> Thanks for example. Yes, that is also extension specific test, which i planned to test it in our extension test internally.

You probably also want a basic test in OSS code to prevent someone cleaning up unused code in the future.

File: source/extensions/transport_sockets/tls/cert_validator/cert_validator.h
Comment (ggreenway) [inline source/extensions/transport_sockets/tls/cert_validator/cert_validator.h:51]: Would it make sense to store the `connectionInfoProvider()` here instead? Then if another user needed the remote address it would be available (as well as a few other things).

Comment (tyxia) [reply source/extensions/transport_sockets/tls/cert_validator/cert_validator.h:51]: @danzh2010 , will quic stack have this info and will it work for quick stack?

@ggreenway  Context: local address was added like common divisor between quic and tcp stack. That is reason we only add what we need.

Comment (ggreenway) [reply source/extensions/transport_sockets/tls/cert_validator/cert_validator.h:51]: Oh, that makes sense. I'm fine leaving it as just local_address then.

**Comment by tyxia** [comment]
@ggreenway and @danzh2010 . We have validated this local address usage in our internal extension. 
Existing ssl_integration_test has already hit this path multiple times. In other words, it is a safe change with test covered by ssl_integration_test.  I think it is a bit overkill to add a test validator extension just for a few lines of change here that are safe. But I definitely can add one if that is strongly preferred. If so, I hope to add it in a followup pr so our internal development can be unblocked.  

Please let me know what do you think.

**Comment by yanavlasov** [comment]
I think I'm with @danzh2010 on adding the test. It is odd to add new functionality without any test coverage. It should at least validate that the plumbing is working correctly, and the value is observed in the extension.

/wait-any

**Comment by tyxia** [comment]
> I think I'm with @danzh2010 on adding the test. It is odd to add new functionality without any test coverage. It should at least validate that the plumbing is working correctly, and the value is observed in the extension.
> 
> /wait-any

The plumbing and the value has been already validated in our internal extension's integration test.. Existing `ssl_integration_test` also validate that this PR is safe. The change here is in callback that is deeply nested in the code which i think integration test with custom extension or a whole new extension is only way to validate it, which I think it is overkill here

I am ok with adding the test validator if it is strongly preferred. But I hope to do it in a follow-up PR to unblock

**Comment by tyxia** [comment]
Moreover, IMHO, the value of integration test that requires whole test validator is quite limited.
1. The only thing we can probably test there is this localAddress is available to be used. i.e., it is not nullptr.
2. We can not test the usage of this local address because there is no validator in OSS that use it and there is no remote validation server in OSS requires it.

Regarding 1: I can change the code to something like below to further prove its accessibility, which i confirmed with `ssl_integration_test` . But i chosen a conservative approach because i feel it is a bit disruptive to existing code.
```
  ASSERT(callbacks != nullptr);
  ASSERT(callbacks->connection().connectionInfoProvider().localAddress() != nullptr);
```

If you feel the code example above is good or any other testing value can be added with a dummy test validator, please let me know.

**Comment by tyxia** [comment]
@yanavlasov @danzh2010 @ggreenway , i have added the test in `ssl_socket_test` to validate the access method that is used in caller site.  Plumbing here is simple as function argument passing. So I think callee site/plumbing is fine as long as the argument passed is available and accessible on the caller site.

PTAL. Thanks

**Comment by yanavlasov** [comment]
Can you add checks for expected context to the test/extensions/transport_sockets/tls/cert_validator/timed_cert_validator.cc ?

/wait-any

**Comment by phlax** [comment]
please hold off pushing CI for the next few hours while we push through release CI

**Comment by tyxia** [comment]
It was my mistake that i didn't realized `timed_cert_validator` is a test validator until I saw its path from test/extension. (name of the file is not that obvious). I have added the integration test to validate the local address. @danzh2010 @yanavlasov @ggreenway  PTAL. Thanks!

**Review by yanavlasov — APPROVED** [review]

File: source/extensions/transport_sockets/tls/context_impl.cc
Comment (danzh2010) [inline source/extensions/transport_sockets/tls/context_impl.cc:521]: IIRC static_cast won't return nullptr upon failure. Maybe use dynamic_cast?

File: test/extensions/transport_sockets/tls/cert_validator/timed_cert_validator.cc
Comment (danzh2010) [inline test/extensions/transport_sockets/tls/cert_validator/timed_cert_validator.cc:25]: Should it also expect validation_context.local_address is not nullptr in this case?

File: test/extensions/transport_sockets/tls/integration/ssl_integration_test.cc
Comment (danzh2010) [inline test/extensions/transport_sockets/tls/integration/ssl_integration_test.cc:494]: Nice test!

File: source/extensions/transport_sockets/tls/context_impl.cc
Comment (tyxia) [reply source/extensions/transport_sockets/tls/context_impl.cc:521]: dynamic_cast for non-polymorphic types are not allowed. `SSL_get_ex_data` returns void pointer. So dynamic_cast can not be used here.

File: test/extensions/transport_sockets/tls/cert_validator/timed_cert_validator.cc
Comment (tyxia) [reply test/extensions/transport_sockets/tls/cert_validator/timed_cert_validator.cc:25]: Those two are a bit orthogonal in theory and local_address is added as optional field (i.e. nullptr).
When callbacks in `context_impl.h` is nullptr for whatever reasons I haven't run into, the disconnected situation happens. In other words, if such case happens in the future(which is possible that callbacks is nullptr because it is returned by `CRYPTO_get_ex_data`), the test will fail if we tie those two here. 

So on safe side, i check nullptr.

File: source/extensions/transport_sockets/tls/cert_validator/cert_validator.h
Comment (tyxia) [reply source/extensions/transport_sockets/tls/cert_validator/cert_validator.h]: Done

File: source/extensions/transport_sockets/tls/context_impl.cc
Comment (danzh2010) [reply source/extensions/transport_sockets/tls/context_impl.cc:521]: I see. In which case can callbacks be nullptr?

File: test/extensions/transport_sockets/tls/cert_validator/timed_cert_validator.cc
Comment (danzh2010) [reply test/extensions/transport_sockets/tls/cert_validator/timed_cert_validator.cc:25]: What if local_address is not at all?

File: source/extensions/transport_sockets/tls/context_impl.cc
Comment (tyxia) [reply source/extensions/transport_sockets/tls/context_impl.cc:521]: Here is how we get callback:
```
void *SSL_get_ex_data(const SSL *ssl, int idx) {
  return CRYPTO_get_ex_data(&ssl->ex_data, idx);
}
```

Based on boringssl implementation, it could be null [here](https://github.com/google/boringssl/blob/e4acd6cb568214b1c7db4e59ce54ea2e1deae1f5/crypto/ex_data.c#L200)

File: test/extensions/transport_sockets/tls/cert_validator/timed_cert_validator.cc
Comment (tyxia) [reply test/extensions/transport_sockets/tls/cert_validator/timed_cert_validator.cc:25]: Please see my reply above. 
I can change it to assert or check it here as localAddress is probably not nullptr in practice. But the goal of this PR is adding **optional** field that is only required by specific extension. It is also the reason I changed it to sharePtr pointer based on previous comments.

So here is really the nits I don't have preference. But I want to choose the safest approach as it is an optional field .

Comment (danzh2010) [reply test/extensions/transport_sockets/tls/cert_validator/timed_cert_validator.cc:25]: Does this test prevent regression if someone revert your change in ContextImpl?

Comment (tyxia) [reply test/extensions/transport_sockets/tls/cert_validator/timed_cert_validator.cc]: well, if people want to revert the change, they can just remove the test as well.
but i don't mind to add a small blocker by adding an assert.

**Review by danzh2010 — APPROVED** [review]
LGTM

**Comment by tyxia** [comment]
@ggreenway gentle ping; the other two reviewers have approved.

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**
